openmc/geometry.py

```diff
@@ -0,0 +1,610 @@
+                cell_name = cell_name.lower()
+
+            if cell_name == name:
+                cells.add(cell)
+            elif not matching and name in cell_name:
+                cells.add(cell)
+
+        return sorted(cells, key=lambda x: x.id)
+
+    def get_cells_by_fill_name(self, name, case_sensitive=False, matching=False):
+        """Return a list of cells with fills with matching names.
+
+        Parameters
+        ----------
+        name : str
+            The name to match
+        case_sensitive : bool
+            Whether to distinguish upper and lower case letters in each
+            cell's name (default is False)
+        matching : bool
+            Whether the names must match completely (default is False)
+
+        Returns
+        -------
+        list of openmc.Cell
+            Cells with fills matching the queried name
+
+        """
+
+        if not case_sensitive:
+            name = name.lower()
+
+        cells = set()
+
+        for cell in self.get_all_cells().values():
+            names = []
+            if cell.fill_type in ('material', 'universe', 'lattice'):
+                names.append(cell.fill.name)
+            elif cell.fill_type == 'distribmat':
+                for mat in cell.fill:
+                    if mat is not None:
+                        names.append(mat.name)
+
+            for fill_name in names:
+                if not case_sensitive:
+                    fill_name = fill_name.lower()
+
+                if fill_name == name:
+                    cells.add(cell)
+                elif not matching and name in fill_name:
+                    cells.add(cell)
+
+        return sorted(cells, key=lambda x: x.id)
+
+    def get_universes_by_name(self, name, case_sensitive=False, matching=False):
+        """Return a list of universes with matching names.
+
+        Parameters
+        ----------
+        name : str
+            The name to match
+        case_sensitive : bool
+            Whether to distinguish upper and lower case letters in each
+            universe's name (default is False)
+        matching : bool
+            Whether the names must match completely (default is False)
+
+        Returns
+        -------
+        list of openmc.Universe
+            Universes matching the queried name
+
+        """
+
+        if not case_sensitive:
+            name = name.lower()
+
+        all_universes = self.get_all_universes().values()
+        universes = set()
+
+        for universe in all_universes:
+            universe_name = universe.name
+            if not case_sensitive:
+                universe_name = universe_name.lower()
+
+            if universe_name == name:
+                universes.add(universe)
+            elif not matching and name in universe_name:
+                universes.add(universe)
+
+        return sorted(universes, key=lambda x: x.id)
+
+    def get_lattices_by_name(self, name, case_sensitive=False, matching=False):
```

Comment:
`get_XXX_by_name` essentially is duplicated code across Cells, Lattices, Universe, and Material. This could all be compressed into one function if all of these derive from a same base class (which just sets the name) or using reflection. This would be a larger change but worth considering for the future.